### PR TITLE
881/multivalue

### PR DESCRIPTION
**Pull request by cornelinux**
This allows custom multi value attributes for every resolver.
This way privacyIDEA can return all groups (memberOf) of a user.

@nomennesc-io you are welcome to comment on this!
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/privacyidea/privacyidea/pull/882%23issuecomment-353627935%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23issuecomment-353628474%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23issuecomment-353960363%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23issuecomment-354070177%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23discussion_r159875445%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23discussion_r160020186%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23discussion_r160020483%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23issuecomment-355901040%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23issuecomment-355902281%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23issuecomment-355956555%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23issuecomment-353627935%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%23%20%5BCodecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882%3Fsrc%3Dpr%26el%3Dh1%29%20Report%5Cn%3E%20Merging%20%5B%23882%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882%3Fsrc%3Dpr%26el%3Ddesc%29%20into%20%5Bmaster%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/commit/d006f7edbb35c686d530d03a258e75f1a7553c8e%3Fsrc%3Dpr%26el%3Ddesc%29%20will%20%2A%2Aincrease%2A%2A%20coverage%20by%20%60%3C.01%25%60.%5Cn%3E%20The%20diff%20coverage%20is%20%60100%25%60.%5Cn%5Cn%5B%21%5BImpacted%20file%20tree%20graph%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882/graphs/tree.svg%3Fwidth%3D650%26height%3D150%26src%3Dpr%26token%3D7nHLZki60B%29%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882%3Fsrc%3Dpr%26el%3Dtree%29%5Cn%5Cn%60%60%60diff%5Cn%40%40%20%20%20%20%20%20%20%20%20%20%20%20Coverage%20Diff%20%20%20%20%20%20%20%20%20%20%20%20%20%40%40%5Cn%23%23%20%20%20%20%20%20%20%20%20%20%20master%20%20%20%20%20%23882%20%20%20%20%20%20%2B/-%20%20%20%23%23%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%2B%20Coverage%20%20%2095.34%25%20%20%2095.35%25%20%20%20%2B%3C.01%25%20%20%20%20%20%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%20%20Files%20%20%20%20%20%20%20%20%20126%20%20%20%20%20%20126%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5Cn%20%20Lines%20%20%20%20%20%20%2015697%20%20%20%2015699%20%20%20%20%20%20%20%2B2%20%20%20%20%20%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%2B%20Hits%20%20%20%20%20%20%20%2014967%20%20%20%2014969%20%20%20%20%20%20%20%2B2%20%20%20%20%20%5Cn%20%20Misses%20%20%20%20%20%20%20%20730%20%20%20%20%20%20730%5Cn%60%60%60%5Cn%5Cn%5Cn%7C%20%5BImpacted%20Files%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882%3Fsrc%3Dpr%26el%3Dtree%29%20%7C%20Coverage%20%5Cu0394%20%7C%20%7C%5Cn%7C---%7C---%7C---%7C%5Cn%7C%20%5Bprivacyidea/lib/resolvers/LDAPIdResolver.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL3Jlc29sdmVycy9MREFQSWRSZXNvbHZlci5weQ%3D%3D%29%20%7C%20%6093.08%25%20%3C100%25%3E%20%28%2B0.03%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%5Cn------%5Cn%5Cn%5BContinue%20to%20review%20full%20report%20at%20Codecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882%3Fsrc%3Dpr%26el%3Dcontinue%29.%5Cn%3E%20%2A%2ALegend%2A%2A%20-%20%5BClick%20here%20to%20learn%20more%5D%28https%3A//docs.codecov.io/docs/codecov-delta%29%5Cn%3E%20%60%5Cu0394%20%3D%20absolute%20%3Crelative%3E%20%28impact%29%60%2C%20%60%5Cu00f8%20%3D%20not%20affected%60%2C%20%60%3F%20%3D%20missing%20data%60%5Cn%3E%20Powered%20by%20%5BCodecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882%3Fsrc%3Dpr%26el%3Dfooter%29.%20Last%20update%20%5Bd006f7e...f0ed05b%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/882%3Fsrc%3Dpr%26el%3Dlastupdated%29.%20Read%20the%20%5Bcomment%20docs%5D%28https%3A//docs.codecov.io/docs/pull-request-comments%29.%5Cn%22%2C%20%22created_at%22%3A%20%222017-12-22T16%3A10%3A15Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars0.githubusercontent.com/u/8655789%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/codecov-io%22%7D%7D%2C%20%7B%22body%22%3A%20%22Looks%20good%21%20I%27m%20wondering%20whether%20member%20and%20memberOf%20should%20be%20included%20by%20default%20next%20to%20mobile.%20I%20understand%20the%20backward%20compatibility%20concern%2C%20but%20I%27m%20having%20a%20hard%20time%20imagining%20anyone%20using%20membership%20information%20taken%20from%20LDAP%20and%20deliberately%20wanting%20to%20use%20only%20the%20one%20group%20that%20quasi-randomly%20popped%20up%20first%20in%20the%20membership-list..%5Cr%5CnPerhaps%20understanding%20the%20rationale%20behind%20mobile%20being%20multi-value%20would%20help%20in%20deciding%3F%20%28I%20have%20no%20idea%20why%20that%20was%20hardcoded%20as%20a%20multi-value%20attribute%29%5Cr%5CnThat%20said%2C%20I%27m%20very%20happy%20to%20see%20this%20as%20a%20configurable%20option%2C%20thanks%21%22%2C%20%22created_at%22%3A%20%222017-12-22T16%3A13%3A08Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/22728411%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/nomennesc-io%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3Ashipit%3A%22%2C%20%22created_at%22%3A%20%222017-12-26T11%3A55%3A30Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/22728411%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/nomennesc-io%22%7D%7D%2C%20%7B%22body%22%3A%20%22%40nomennesc-io%20the%20%60%60mobile%60%60%20attribute%20is%20the%20one%20attribute%20which%20has%20a%20practical%20use.%20%28Well%2C%20we%20could%20argue%20email%20has%2C%20too%29.%20But%20honestly%20I%20absolutely%20do%20not%20like%20email%20as%20a%20means%20to%20send%20OTP%20values%21%20I%20also%20do%20not%20like%20SMS%20that%20much%20either.%5Cr%5CnThe%20%60%60mobile%60%60%20attribute%20is%20used%20to%20display%20default%20values%2C%20or%20suggest%20values%20for%20mobile%20numbers%2C%20when%20enrolling%20an%20SMS%20token%20for%20the%20user.%5Cr%5Cn%5Cr%5CnI%20think%20if%20someone%20is%20that%20deep%20into%20privacyIDEA%20like%20when%20mapping%20arbitrary%20attributes%2C%20such%20user/admin%20might%20as%20well%20decide%2C%20if%20he%20wants%20to%20have%20a%20single%20value%20or%20a%20multi%20value%20attribute.%5Cr%5Cn%5Cr%5CnThanks%20for%20your%20feedback%21%22%2C%20%22created_at%22%3A%20%222017-12-27T07%3A52%3A00Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%2C%20%7B%22body%22%3A%20%22Documentation%20looks%20fine%20for%20me%20now%21%5Cr%5CnI%20stumbled%20upon%20one%20small%20remaining%20issue%3A%20If%20I%20set%20some%20value%20for%20%60%60Multivalue%20Attributes%60%60%2C%20save%20it%2C%20then%20delete%20the%20value%20and%20save%20again%2C%20authentication%20fails%20with%5Cr%5Cn%60%60%60%5Cr%5Cn...%20%20File%20%5C%22/home/fred/privacyidea/privacyidea/privacyidea/lib/usercache.py%5C%22%2C%20line%2063%2C%20in%20cache_wrapper%5Cr%5Cn%20%20%20%20return%20wrapped_function%28%2Aargs%2C%20%2A%2Akwds%29%5Cr%5Cn%20%20File%20%5C%22/home/fred/privacyidea/privacyidea/privacyidea/lib/user.py%5C%22%2C%20line%20692%2C%20in%20get_username%5Cr%5Cn%20%20%20%20username%20%3D%20y.getUsername%28userid%29%5Cr%5Cn%20%20File%20%5C%22/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py%5C%22%2C%20line%20464%2C%20in%20getUsername%5Cr%5Cn%20%20%20%20info%20%3D%20self.getUserInfo%28user_id%29%5Cr%5Cn%20%20File%20%5C%22/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py%5C%22%2C%20line%20168%2C%20in%20cache_wrapper%5Cr%5Cn%20%20%20%20f_result%20%3D%20func%28self%2C%20%2Aargs%2C%20%2A%2Akwds%29%5Cr%5Cn%20%20File%20%5C%22/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py%5C%22%2C%20line%20427%2C%20in%20getUserInfo%5Cr%5Cn%20%20%20%20ret%20%3D%20self._ldap_attributes_to_user_object%28attributes%29%5Cr%5Cn%20%20File%20%5C%22/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py%5C%22%2C%20line%20445%2C%20in%20_ldap_attributes_to_user_object%5Cr%5Cn%20%20%20%20elif%20type%28ldap_v%29%20%3D%3D%20list%20and%20map_k%20not%20in%20self.multivalueattributes%3A%5Cr%5CnTypeError%3A%20argument%20of%20type%20%27NoneType%27%20is%20not%20iterable%5Cr%5Cn%60%60%60%5Cr%5Cnbecause%20here%3A%5Cr%5Cn%60%60%60python%5Cr%5Cnmultivalueattributes%20%3D%20config.get%28%5C%22MULTIVALUEATTRIBUTES%5C%22%2C%20%27%5B%5C%22mobile%5C%22%5D%27%29%5Cr%5Cnself.multivalueattributes%20%3D%20yaml.safe_load%28multivalueattributes%29%5Cr%5Cn%60%60%60%5Cr%5Cnthe%20MULTIVALUEATTRIBUTES%20value%20is%20the%20empty%20string%2C%20which%20%60%60yaml.safe_load%60%60%20parses%20to%20%60%60None%60%60%20%3A-/%22%2C%20%22created_at%22%3A%20%222018-01-08T07%3A52%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/28243%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/fredreichbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%2C%20we%20need%20to%20do%5Cn%5Cnmultivalueattributes%20%3D%20config.get%28%5C%22MULTIVALUEATTRIBUTES%5C%22%29%20or%5Cu00a0%27%5B%5C%22mobile%5C%22%5D%27%5Cn%5Cn%5Cn%5Cnnull%22%2C%20%22created_at%22%3A%20%222018-01-08T08%3A01%3A19Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%2C%20%7B%22body%22%3A%20%22Closes%20%23882%20%22%2C%20%22created_at%22%3A%20%222018-01-08T12%3A42%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20f0ed05b631b863231329a3d13ebc6b17f640723b%20privacyidea/static/components/config/views/config.resolvers.ldap.html%206%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23discussion_r160020186%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Fixed%20it%20in%206369f5c0a847a9061e4c6433cd3b359c8721ed79%22%2C%20%22created_at%22%3A%20%222018-01-06T07%3A40%3A49Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/static/components/config/views/config.resolvers.ldap.html%3AL264-279%22%7D%2C%20%22Pull%20f0ed05b631b863231329a3d13ebc6b17f640723b%20doc/configuration/useridresolvers.rst%209%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/882%23discussion_r159875445%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20think%20the%20first%20sentence%20is%20a%20bit%20unclear%3A%20For%20me%2C%20it%20reads%20like%20MULTIVALUEATTRIBUTES%20contains%20a%20list%20of%20LDAP%20attribute%20names%20%28i.e.%20%60%60mobile%60%60%20and%20%60%60memberOf%60%60%20are%20attributes%20of%20the%20user%20object%20in%20the%20LDAP%20tree%29.%20However%2C%20actually%2C%20MULTIVALUEATTRIBUTES%20seems%20to%20contain%20privacyIDEA%20user%20info%20keys%2C%20i.e.%20the%20keys%20the%20LDAP%20attributes%20are%20mapped%20to.%5Cr%5Cn%5Cr%5CnAs%20an%20example%2C%20if%20I%20configure%20my%20mapping%20like%20this%3A%5Cr%5Cn%60%60%60%5Cr%5Cn%7B%5C%22group%5C%22%3A%20%5C%22memberOf%5C%22%2C%20...%7D%5Cr%5Cn%60%60%60%5Cr%5CnI%20actually%20need%20to%20add%20%60%60group%60%60%2C%20not%20%60%60memberOf%60%60%2C%20to%20MULTIVALUEATTRIBUTES.%20I%20think%20the%20documentation%20could%20make%20this%20a%20bit%20clearer.%5Cr%5Cn%5Cr%5CnAlso%2C%20%5C%22The%20setting%5C%22%20should%20be%20%5C%22The%20default%20setting%5C%22%20and%20there%20is%20a%20%5C%22user%20user%5C%22%20at%20the%20end%20%3A%29%5Cr%5Cn%20%20%22%2C%20%22created_at%22%3A%20%222018-01-05T13%3A29%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/28243%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/fredreichbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22fixed%20in%20ee6e058f0335d217b33e2b637affa127dff53300%22%2C%20%22created_at%22%3A%20%222018-01-06T08%3A01%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20doc/configuration/useridresolvers.rst%3AL144-156%22%7D%7D%2C%20%22approved%22%3A%20%7B%22https%3A//github.com/nomennesc-io%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/22728411%3Fv%3D4%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>

<a href='https://github.com/nomennesc-io'></a>

- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/882#issuecomment-353627935'>General Comment</a></b>
- <a href='https://github.com/codecov-io'></a> # [Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/882?src=pr&el=h1) Report
> Merging [#882](https://codecov.io/gh/privacyidea/privacyidea/pull/882?src=pr&el=desc) into [master](https://codecov.io/gh/privacyidea/privacyidea/commit/d006f7edbb35c686d530d03a258e75f1a7553c8e?src=pr&el=desc) will **increase** coverage by `<.01%`.
> The diff coverage is `100%`.
[Impacted file tree graph](https://codecov.io/gh/privacyidea/privacyidea/pull/882?src=pr&el=tree)
```diff
@@            Coverage Diff             @@
##           master     #882      +/-   ##
==========================================
+ Coverage   95.34%   95.35%   +<.01%
==========================================
Files         126      126
Lines       15697    15699       +2
==========================================
+ Hits        14967    14969       +2
Misses        730      730
```
| [Impacted Files](https://codecov.io/gh/privacyidea/privacyidea/pull/882?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [privacyidea/lib/resolvers/LDAPIdResolver.py](https://codecov.io/gh/privacyidea/privacyidea/pull/882/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL3Jlc29sdmVycy9MREFQSWRSZXNvbHZlci5weQ==) | `93.08% <100%> (+0.03%)` | :arrow_up: |
------
[Continue to review full report at Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/882?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/882?src=pr&el=footer). Last update [d006f7e...f0ed05b](https://codecov.io/gh/privacyidea/privacyidea/pull/882?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).
- <a href='https://github.com/nomennesc-io'></a> Looks good! I'm wondering whether member and memberOf should be included by default next to mobile. I understand the backward compatibility concern, but I'm having a hard time imagining anyone using membership information taken from LDAP and deliberately wanting to use only the one group that quasi-randomly popped up first in the membership-list..
Perhaps understanding the rationale behind mobile being multi-value would help in deciding? (I have no idea why that was hardcoded as a multi-value attribute)
That said, I'm very happy to see this as a configurable option, thanks!
- <a href='https://github.com/nomennesc-io'></a> :shipit:
- <a href='https://github.com/cornelinux'></a> @nomennesc-io the ``mobile`` attribute is the one attribute which has a practical use. (Well, we could argue email has, too). But honestly I absolutely do not like email as a means to send OTP values! I also do not like SMS that much either.
The ``mobile`` attribute is used to display default values, or suggest values for mobile numbers, when enrolling an SMS token for the user.
I think if someone is that deep into privacyIDEA like when mapping arbitrary attributes, such user/admin might as well decide, if he wants to have a single value or a multi value attribute.
Thanks for your feedback!
- <a href='https://github.com/fredreichbier'></a> Documentation looks fine for me now!
I stumbled upon one small remaining issue: If I set some value for ``Multivalue Attributes``, save it, then delete the value and save again, authentication fails with
```
...  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/usercache.py", line 63, in cache_wrapper
return wrapped_function(*args, **kwds)
File "/home/fred/privacyidea/privacyidea/privacyidea/lib/user.py", line 692, in get_username
username = y.getUsername(userid)
File "/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py", line 464, in getUsername
info = self.getUserInfo(user_id)
File "/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py", line 168, in cache_wrapper
f_result = func(self, *args, **kwds)
File "/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py", line 427, in getUserInfo
ret = self._ldap_attributes_to_user_object(attributes)
File "/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py", line 445, in _ldap_attributes_to_user_object
elif type(ldap_v) == list and map_k not in self.multivalueattributes:
TypeError: argument of type 'NoneType' is not iterable
```
because here:
```python
multivalueattributes = config.get("MULTIVALUEATTRIBUTES", '["mobile"]')
self.multivalueattributes = yaml.safe_load(multivalueattributes)
```
the MULTIVALUEATTRIBUTES value is the empty string, which ``yaml.safe_load`` parses to ``None`` :-/
- <a href='https://github.com/cornelinux'></a> Ok, we need to do
multivalueattributes = config.get("MULTIVALUEATTRIBUTES") or '["mobile"]'
null
- <a href='https://github.com/cornelinux'></a> Closes #882
- [x] <a href='#crh-comment-Pull f0ed05b631b863231329a3d13ebc6b17f640723b doc/configuration/useridresolvers.rst 9'></a> &nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/882#discussion_r159875445'>File: doc/configuration/useridresolvers.rst:L144-156</a></b>
- <a href='https://github.com/fredreichbier'></a> I think the first sentence is a bit unclear: For me, it reads like MULTIVALUEATTRIBUTES contains a list of LDAP attribute names (i.e. ``mobile`` and ``memberOf`` are attributes of the user object in the LDAP tree). However, actually, MULTIVALUEATTRIBUTES seems to contain privacyIDEA user info keys, i.e. the keys the LDAP attributes are mapped to.
As an example, if I configure my mapping like this:
```
{"group": "memberOf", ...}
```
I actually need to add ``group``, not ``memberOf``, to MULTIVALUEATTRIBUTES. I think the documentation could make this a bit clearer.
Also, "The setting" should be "The default setting" and there is a "user user" at the end :)
- <a href='https://github.com/cornelinux'></a> fixed in ee6e058f0335d217b33e2b637affa127dff53300
- [x] <a href='#crh-comment-Pull f0ed05b631b863231329a3d13ebc6b17f640723b privacyidea/static/components/config/views/config.resolvers.ldap.html 6'></a> &nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/882#discussion_r160020186'>File: privacyidea/static/components/config/views/config.resolvers.ldap.html:L264-279</a></b>
- <a href='https://github.com/cornelinux'></a> Fixed it in 6369f5c0a847a9061e4c6433cd3b359c8721ed79

<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/882?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/882?approve=1'></a>&nbsp;<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/882?approve=0'></a>&nbsp;<a href='https://github.com/privacyidea/privacyidea/pull/882'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by cornelinux** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/privacyidea/privacyidea/pull/882'>Refresh</a>.<a href='#crh-update-none'></a>

**Comment by nomennesc-io** [comment]
Looks good! I'm wondering whether member and memberOf should be included by default next to mobile. I understand the backward compatibility concern, but I'm having a hard time imagining anyone using membership information taken from LDAP and deliberately wanting to use only the one group that quasi-randomly popped up first in the membership-list..
Perhaps understanding the rationale behind mobile being multi-value would help in deciding? (I have no idea why that was hardcoded as a multi-value attribute)
That said, I'm very happy to see this as a configurable option, thanks!

**Comment by nomennesc-io** [comment]
:shipit:

**Comment by cornelinux** [comment]
@nomennesc-io the ``mobile`` attribute is the one attribute which has a practical use. (Well, we could argue email has, too). But honestly I absolutely do not like email as a means to send OTP values! I also do not like SMS that much either.
The ``mobile`` attribute is used to display default values, or suggest values for mobile numbers, when enrolling an SMS token for the user.

I think if someone is that deep into privacyIDEA like when mapping arbitrary attributes, such user/admin might as well decide, if he wants to have a single value or a multi value attribute.

Thanks for your feedback!

File: privacyidea/static/components/config/views/config.resolvers.ldap.html [hunk 1]
```diff
@@ -264,6 +264,15 @@
                         />
             </div>
         </div>
+        <div class="form-group">
+            <label for="multivalueattributes" class="col-sm-3 control-label"
+                translate>Mutlivalue Attributes</label>
```
Comment (fredreichbier) [inline privacyidea/static/components/config/views/config.resolvers.ldap.html]: *Multivalue :)

File: doc/configuration/useridresolvers.rst [hunk 2]
```diff
@@ -144,6 +144,12 @@ attribute mapping with a key, you make up and the LDAP attribute like:
 "homeDirectory" and "objectGUID" being the attributes in the LDAP directory
 and "homedir" and "studentID" the keys returned in a SAML authentication
 request.
+
+The ``MULTIVALUEATTRIBUTES`` config value can be used to specify a list of
+LDAP user attributes, that should return a list of values. The setting is the
+"mobile" attribute, i.e. privacyIDEA will return a list of mobile telephone numbers.
+E.g. you can set this to ``["mobile", "memberOf"]`` and then it would also return the
+list of groups of the user user.
```
Comment (fredreichbier) [inline doc/configuration/useridresolvers.rst]: I think the first sentence is a bit unclear: For me, it reads like MULTIVALUEATTRIBUTES contains a list of LDAP attribute names (i.e. ``mobile`` and ``memberOf`` are attributes of the user object in the LDAP tree). However, actually, MULTIVALUEATTRIBUTES seems to contain privacyIDEA user info keys, i.e. the keys the LDAP attributes are mapped to.

As an example, if I configure my mapping like this:
```
{"group": "memberOf", ...}
```
I actually need to add ``group``, not ``memberOf``, to MULTIVALUEATTRIBUTES. I think the documentation could make this a bit clearer.

Also, "The setting" should be "The default setting" and there is a "user user" at the end :)

File: privacyidea/static/components/config/views/config.resolvers.ldap.html (hunk 1 above)
Comment (cornelinux) [reply privacyidea/static/components/config/views/config.resolvers.ldap.html]: Fixed it in 6369f5c0a847a9061e4c6433cd3b359c8721ed79

File: doc/configuration/useridresolvers.rst (hunk 2 above)
Comment (cornelinux) [reply doc/configuration/useridresolvers.rst]: fixed in ee6e058f0335d217b33e2b637affa127dff53300

**Comment by fredreichbier** [comment]
Documentation looks fine for me now!
I stumbled upon one small remaining issue: If I set some value for ``Multivalue Attributes``, save it, then delete the value and save again, authentication fails with
```
...  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/usercache.py", line 63, in cache_wrapper
    return wrapped_function(*args, **kwds)
  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/user.py", line 692, in get_username
    username = y.getUsername(userid)
  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py", line 464, in getUsername
    info = self.getUserInfo(user_id)
  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py", line 168, in cache_wrapper
    f_result = func(self, *args, **kwds)
  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py", line 427, in getUserInfo
    ret = self._ldap_attributes_to_user_object(attributes)
  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/resolvers/LDAPIdResolver.py", line 445, in _ldap_attributes_to_user_object
    elif type(ldap_v) == list and map_k not in self.multivalueattributes:
TypeError: argument of type 'NoneType' is not iterable
```
because here:
```python
multivalueattributes = config.get("MULTIVALUEATTRIBUTES", '["mobile"]')
self.multivalueattributes = yaml.safe_load(multivalueattributes)
```
the MULTIVALUEATTRIBUTES value is the empty string, which ``yaml.safe_load`` parses to ``None`` :-/

**Comment by cornelinux** [comment]
Ok, we need to do

multivalueattributes = config.get("MULTIVALUEATTRIBUTES") or '["mobile"]'



null

**Comment by cornelinux** [comment]
Closes #882

**Pull request merged**